corehq/apps/hqwebapp/doc_lookup.py

```diff
@@ -0,0 +1,178 @@
+from collections import OrderedDict, defaultdict, namedtuple
+
+import attr
+from couchdbkit import ResourceNotFound
+from django.core.exceptions import ObjectDoesNotExist
+from memoized import memoized
+
+from corehq.apps.locations.models import SQLLocation
+from corehq.form_processor.exceptions import CaseNotFound, XFormNotFound
+from corehq.form_processor.models import CommCareCaseSQL, XFormInstanceSQL
+from corehq.form_processor.serializers import (
+    CommCareCaseSQLRawDocSerializer,
+    XFormInstanceSQLRawDocSerializer,
+)
+from corehq.util.couchdb_management import couch_config
+
+DbResult = namedtuple('DbResult', 'dbname result status')
+
+
+@attr.s
+class LookupResult:
+    doc_id = attr.ib()
+    doc = attr.ib()
+    doc_type = attr.ib()
+    domain = attr.ib()
+    dbname = attr.ib()
+    data_serializer = attr.ib()
+
+    def asdict(self):
+        res = attr.asdict(self)
+        res.pop("data_serializer")
+        return res
+
+    def get_serialized_doc(self):
+        if not self.data_serializer:
+            return self.doc
+        return self.data_serializer(self.doc)
+
+
+def lookup_doc_id(doc_id):
+    """Look up a document ID
+
+    :param doc_id: ID of document to look for
+    :returns: LookupResult or None
+    """
+    result, _ = lookup_id_in_databases(doc_id, get_databases().values())
+    return result
+
+
+def lookup_id_in_databases(doc_id, dbs, find_first=True):
+    """Look up a document ID
+
+    :param doc_id: ID of document to look for
+    :param dbs: list of databases to look in
+    :param find_first: Set to False to look in all databases even after a match is found.
+    :returns: Tuple of (LookupResult or None, list of DbResult objects for each DB searched)
+    """
+    STATUSES = defaultdict(lambda: 'warning', {
+        'missing': 'default',
+        'deleted': 'danger',
+    })
+
+    db_results = []
+    response = None
+    for db in dbs:
+        try:
+            db_response = db.get_context(doc_id)
+            if not response:
+                response = db_response
+        except ResourceNotFound as e:
+            db_results.append(DbResult(db.dbname, str(e), STATUSES[str(e)]))
+        else:
+            db_results.append(DbResult(db.dbname, 'found', 'success'))
+            if find_first:
+                break
+
+    return response, db_results
+
+
+@memoized
+def get_databases():
+    """Return an ordered dict of (dbname: database). The order is
+    according to search preference, the first DB to contain a document
+    should be assumed to be the authoritative one."""
+    sql_dbs = [
```

Comment:
Out of scope for this PR, but I've often wished we had a way to lookup blobs by id. Seems like it would be pretty easy to add to this list.